### ISSUE TITLE
Downstream PR preview creation and build slowness/timeouts

### ISSUE THREAD

**Issue by victorlin** [nextstrain/auspice#1709]
## Context

When working on UI changes, it's useful to previews the changes on nextstrain.org. Currently, [the process](https://github.com/nextstrain/auspice/blob/8377c4c47003468dab5b34cc02ab2d06eb952ca1/DEV_DOCS.md#test-on-downstream-repositories) is simple but it takes a long time, at least 20 minutes: >10m to make the PR, then >10m to build the preview version of nextstrain.org. The slowness seems to be mostly due to `npm install` being slow when resolving git references.

## Upstream issue

- https://github.com/npm/cli/issues/4896

## Past discussions

This has been disussed in https://github.com/nextstrain/auspice/pull/1567#issuecomment-1273892375 and https://github.com/nextstrain/auspice/pull/1661#issuecomment-1500667474. I thought it would be fixed with the upgrade to Node.js 16, but looks like it's still taking a while.

## Solutions

- For time to create the downstream PR: #1870
- For time to build the review app on Heroku: https://github.com/nextstrain/nextstrain.org/pull/1041

## Solutions that didn't work / weren't pursued

1. Accept the slowness and do things to fill the void while waiting, such as writing issues like this.
2. Try using `yarn`, maybe it's faster?
3. Do the `npm install` with `https://`
4. Replace `github:` with `https://` in package.json + package-lock.json
5. (in git config on both the workflow that creates the PR and the heroku build env) Use `url."ssh://git@".insteadOf git://` or `url."https://github".insteadOf "git://github"`

**Comment by jameshadfield** [comment]
Recent changes to how nextstrain.org is built & deployed have improved this a lot. `npm` is still slower installing from GitHub cf. the npm registry, but I don't think it's worth switching to `yarn` or similar. Please re-open this issue if it's still frustrating for anyone!

**Comment by victorlin** [comment]
Re-opening this because it is once again taking at least 20 minutes and even [timing out](https://github.com/nextstrain/nextstrain.org/pull/1025#issuecomment-2420582251) on Heroku builds.

**Comment by victorlin** [comment]
Read through [the upstream issue](https://github.com/npm/cli/issues/4896) and added some more possible solutions which I'm trying in #1870.

**Comment by victorlin** [comment]
In https://github.com/nextstrain/nextstrain.org/pull/1039 I tried some different workarounds mentioned in https://github.com/npm/cli/issues/4896 but none seemed to work:

- use `https://` – https://github.com/nextstrain/nextstrain.org/commit/b0f539a491222b7ff1935eef2ffc460f3a3bc85d
- use `git+ssh://git@github.com` – https://github.com/nextstrain/nextstrain.org/commit/2fe815cf7e625b7427e2b1d2463a5714add69df3
- set git config `url."ssh://git@".insteadOf git://` – https://github.com/nextstrain/nextstrain.org/commit/f6ad7a10fb55b8885a8dc8a9bbeddf59c0f9a8bf

I then ran `npm ci` verbosely¹ which revealed where the slowness is happening. Here are the logs from the build that ran on [GitHub](https://github.com/nextstrain/nextstrain.org/actions/runs/11396140104/job/31709485515):

```
2024-10-18T02:07:38.2155165Z npm http fetch GET 200 https://registry.npmjs.org/core-js/-/core-js-3.29.1.tgz 6072ms (cache miss)
2024-10-18T02:07:38.2407069Z npm http fetch GET 200 https://codeload.github.com/nextstrain/auspice/tar.gz/80c8cf0c8946ea10e2c3a0ce3caad0ea2653891f 5464ms (cache miss)
2024-10-18T02:14:20.0421014Z npm info run core-js@3.29.1 postinstall node_modules/core-js node -e "try{require('./postinstall')}catch(e){}"
```

and on [Heroku](https://dashboard.heroku.com/apps/nextstrain-s-nextstrain-hbahjw/activity/builds/a63d1974-7a7f-4765-9827-f9102c31a678) (build is still stuck "in progress", I'll update when it is done/times out):

```
npm http fetch GET 200 https://registry.npmjs.org/core-js/-/core-js-3.29.1.tgz 5152ms (cache miss)
```

Observations:

- The Heroku build is stuck on fetching from `https://codeload.github.com/nextstrain/auspice/tar.gz/80c8cf0c8946ea10e2c3a0ce3caad0ea2653891f`. I have a suspicion that this is related to the ongoing ["Open Source Repository Engagement Data" incident](https://status.heroku.com/incidents/2720). I submitted a support request regarding this.
- The GitHub build spent 7 minutes on `npm info run core-js@3.29.1 postinstall node_modules/core-js node -e "try{require('./postinstall')}catch(e){}"` (or something else if verbose logs are still not telling the full story).

¹ it [took me a while](https://github.com/nextstrain/nextstrain.org/compare/f6ad7a10fb55b8885a8dc8a9bbeddf59c0f9a8bf...e40d81da7ae8cecb49639ea1d895aa3ed2e248ed) to figure this one out since Heroku sets `NPM_CONFIG_LOGLEVEL=error` somewhere...

**Comment by victorlin** [comment]
Heroku support identified that it was timing out on puppeteer's download of Chrome and suggested using `PUPPETEER_SKIP_DOWNLOAD=true` to resolve the issue. I confirmed that the Heroku build now completes in ~5 minutes with that setting in https://github.com/nextstrain/nextstrain.org/commit/20dc6c1537914dd92e8f631634a90368c10f481f.

I wouldn't treat that as a solution though. It's not clear to me why puppeteer is being downloaded in the first place since it is listed as a [dev dependency of Auspice](https://github.com/nextstrain/auspice/blob/8e117aa074a6ee5eb1f6d7fd62b4a671a878faba/package.json#L137). On the surface it seems like it [shouldn't](https://stackoverflow.com/a/34719468) because `NODE_ENV` is set to production on the Heroku build. I'll investigate further after lunch.

**Comment by victorlin** [comment]
This issue has two parts:

1. Time to create the downstream PR
2. Time to build the review app from the downstream PR

I've addressed (1) with #1870 which is ready for review. (2) is tricker with the weirdness I mentioned above. I'll continue to investigate that since these builds are still timing out on Heroku.

**Comment by victorlin** [comment]
Re: Heroku build issues, I think I've found an explanation in [npm docs](https://docs.npmjs.com/cli/v10/using-npm/scripts#life-cycle-scripts):

> NOTE: If a package being installed through git contains a `prepare` script, its `dependencies` and `devDependencies` will be installed, and the prepare script will be run, before the package is packaged and installed.

and a reason for the elusiveness in logs:

> As of `npm@7` these scripts run in the background. To see the output, run with: `--foreground-scripts`.

**Comment by victorlin** [comment]
I just applied the solutions to https://github.com/nextstrain/auspice/pull/1373. Here's the improved timeline:

- [3:43pm] [pushed](https://github.com/nextstrain/auspice/pull/1373#commits-pushed-e8586a4) e2d2658979c135df664313a5d465b5a3cf1478c4...1120581f6707963ab36d57db0d7ec613901fd156
- [3:47pm] preview PR [updated](https://github.com/nextstrain/nextstrain.org/pull/1025#event-14728135367)
- [3:51pm] review app updated 
![image](https://github.com/user-attachments/assets/91de88f1-54bd-42aa-a30f-e08840a702e4)

**Comment by victorlin** [comment]
Fixed by #1870 and https://github.com/nextstrain/nextstrain.org/pull/1041

**Comment by genehack** [comment]
👏
